uds/database/abstract_data_record.py

```diff
@@ -0,0 +1,123 @@
+"""
+Definition of all Data Records types.
+
+Each Data Record contains mapping (translation) of raw data (sequence of bits in diagnostic message payload) to some
+meaningful information (e.g. physical value, text).
+"""
+
+__all__ = ["DataRecordType", "AbstractDataRecord", "DecodedDataRecord"]
+
+from abc import ABC, abstractmethod
+from typing import Tuple, TypedDict, Union
+
+from uds.utilities import ValidatedEnum
+
+DataRecordPhysicalValueAlias = Union[int, float, str, Tuple["DecodedDataRecord", ...]]
+"""Alias of Data Records' physical value."""
+
+
+class DecodedDataRecord(TypedDict):
+    """Structure of decoded Data Record."""
+
+    name: str
+    raw_value: int
+    physical_value: DataRecordPhysicalValueAlias  # noqa: F841
+
+
+class DataRecordType(ValidatedEnum):
+    """All Data Record types."""
+
+    # TODO: fill with following tasks:
+    #  - https://github.com/mdabrowski1990/uds/issues/2
+    #  - https://github.com/mdabrowski1990/uds/issues/6
+    #  - https://github.com/mdabrowski1990/uds/issues/8
+    #  - https://github.com/mdabrowski1990/uds/issues/9
+    #  - https://github.com/mdabrowski1990/uds/issues/10
+
+
+class AbstractDataRecord(ABC):
+    """Common implementation and interface for all Data Records."""
+
+    def __init__(self, name: str) -> None:
+        """
+        Initialize common part for all Data Records.
+
+        :param name: Name to assign to this Data Record.
+
+        :raise TypeError: Provided value of name is not str type.
+        """
+        if not isinstance(name, str):
+            raise TypeError("Provided name is not str type.")
+        self.__name = name.strip()
+
+    @property
+    def name(self) -> str:
+        """Name of this Data Record."""
+        return self.__name
+
+    @property  # noqa: F841
+    @abstractmethod
+    def data_record_type(self) -> DataRecordType:
+        """Type of this Data Record."""
+
+    @property  # noqa: F841
+    @abstractmethod
+    def length(self) -> int:
+        """Get number of bits that this Data Record is stored over."""
+
+    @property  # noqa: F841
+    @abstractmethod
+    def is_reoccurring(self) -> bool:
```

Comment:
In which part of UDS protocol communication could these occurrence-related properties be used?

Reply:
for excample to most of ReadDTCInformation Responses (you do not know how many occurrences of `DTCAndStatusRecord` would be there):
![image](https://github.com/user-attachments/assets/93b339d5-5318-4079-82e4-c604254b3ab5)

Reply:
hmm, but shouldnt the DataRecord represent only the four bytes of a particular DTC (DTCHighByte, DTCMiddleByte, DTCLowByte, and statusOfDTC)? As I understand from your answer to my previous question, it should be an object for each individual part of the UDS message. I suppose that in the DTCAndStatusRecord, we will have multiple separate DTCs, so we should get multiple DataRecords there?

Reply:
DataRecord is just containing information, it is not part of diagnostic message.

Example:
Diagnostic Message 1: [0x59, 0x02, 0xFF, 0x12, 0x34, 0x56, 0xFF]
Diagnostic Message 2: [0x59, 0x02, 0xFF, 0x12, 0x34, 0x56, 0xFF, 0xFE, 0xDC, 0xBA, 0x1F]

Messages database (decoder and encoder for all diagnostic messages) would hold ReadDTCInformation Service (decode and encoder to ReadDTCInformation 0x19 and ReadDTCInformation Reponse 0x59 diagnostic messages), then it will hold ReadDTCInformation Response for presented subfunctions, which would contain DataRecords (information about each entry in this table).

End result (simplified, just to give an idea):
Message 1 -> decode -> `{"SID": 0x59, "reaportType": 0x02, "DTCStatusAvailabilityMask": 0xFF", "DTCAmdStatusRecord": [{"DTC": 0x123456", "DTC Status": 0xFF"}]}`
Message 1 -> decode -> `{"SID": 0x59, "reaportType": 0x02, "DTCStatusAvailabilityMask": 0xFF", "DTCAmdStatusRecord": [{"DTC": 0x123456", "DTC Status": 0xFF"}, {"DTC": 0xFEDCBA", "DTC Status": 0x1F"}]}`

Summary:
- Object with [diagnostic message](https://uds.readthedocs.io/en/stable/pages/user_guide/message.html#uds-message) is not divided nor modified, it stays the same, untouched and it is the only object that contains `payload` of diagnostic messages.
- [databases](https://github.com/mdabrowski1990/uds/issues/11) provides way to decode information form diagnostic messages (look example above) and ways to encode (create) a diagnostic message from physical data (e.g. I know I want to set `Extended` diagnostic session, but I do not know which value it represents, so I just provide value `Extended` and database is supposed to find the raw value and use it to create a diagnostic message payload)
- `DataRecord` class contain generic definition of single raw in presented table
- object of `DataRecord` class contains certain information about just one data entry (e.g. mapping for `reportType`: reportDTCByStatusMask = 0x02; reportSupportedDTCs = 0x0A etc.)
- Every task in [this release](https://github.com/mdabrowski1990/uds/milestone/2) is about extending database capabilities (provide tools for database creation/extension)
- Next milestone is about providing base database (the one purily based on ISO 14229-1), which could later be extended by users